### DISCUSSION TITLE
How to add an offline fallback

### DISCUSSION THREAD

**Discussion by FleetAdmiralJakob** [serwist/serwist#13] [Q&A]
So, if a page is not cached and you are offline.

**Comment by DuCanhGH** [comment]
@FleetAdmiralJakob The built-in support will take a while, but here's an alternative:
```js
import withSerwistInit from "@serwist/next";

const withSerwist = withSerwistInit({
  cacheOnFrontEndNav: true,
  swSrc: "app/sw.ts",
  swDest: "public/sw.js",
  additionalPrecacheEntries: ["/fallback-font.ttf"],
});
```

```ts
installSerwist({
  precacheEntries: self.__SW_MANIFEST,
  skipWaiting: true,
  clientsClaim: true,
  navigationPreload: true,
  runtimeCaching: [
    {
      urlPattern: /^https:\/\/fonts\.(?:gstatic)\.com\/.*/i,
      handler: "CacheFirst",
      options: {
        cacheName: "google-fonts-webfonts",
        expiration: {
          maxEntries: 4,
          maxAgeSeconds: 365 * 24 * 60 * 60, // 365 days
        },
        plugins: [
          {
            async handlerDidError() {
              const fallbackResponse = await caches.match("/fallback-font.ttf", { ignoreSearch: true });
              if (fallbackResponse !== undefined) {
                return fallbackResponse;
              }
              return Response.error();
            },
          },
        ],
      },
    },
    ...defaultCache.filter((entry) => !(entry.options?.cacheName === "google-fonts-webfonts")),
  ],
});
```

**Reply by FleetAdmiralJakob** [reply]
Hi, thank you for your answer. But isn't this just for a font and not for an offline fallback page? 🤔

**Reply by DuCanhGH** [reply]
@FleetAdmiralJakob yes, it is for fonts, but the same applies for pages. Simply copy ["pages-rsc-prefetch", "pages-rsc", and "pages"](https://github.com/serwist/serwist/blob/main/packages/next/src/index.browser.ts#L166-L198) from `@serwist/next` (you may want to update their `urlPattern` a bit so as to not cache wrong routes though), add `plugins.handlerDidError`, filter "pages-rsc-prefetch", "pages-rsc", and "pages" from `defaultCache`, and add "/~offline" to `additionalPrecacheEntries` :)

**Reply by FleetAdmiralJakob** [reply]
OK, thank you, I will try it out. Maybe you can add this to the docs so people find it in future more easily

**Reply by DuCanhGH** [reply]
@FleetAdmiralJakob no need. The feature should be ready by the time the docs is ready for use :)

**Reply by DuCanhGH** [reply]
@FleetAdmiralJakob it has finally been implemented in 8.1.0! Here is how to use it:
```js
// Anything random.
const revision = crypto.randomUUID();

installSerwist({
  precacheEntries: self.__SW_MANIFEST,
  skipWaiting: true,
  clientsClaim: true,
  navigationPreload: true,
  runtimeCaching: defaultCache,
  fallbacks: {
    entries: [
      {
        url: "/~offline",
        revision,
        matcher({ request }) {
          return request.destination === "document";
        },
      },
    ],
  },
});
```
A bit different from the old API, I'd say.

**Reply by FleetAdmiralJakob** [reply]
Ok. Why is this implementation better than the old API?

**Reply by DuCanhGH** [reply]
@FleetAdmiralJakob well, one big reason is that you are not tied down to a fixed set of RequestDestination. In fact, you can have as many fallback entries as possible. Another reason is that it is implemented into `@serwist/sw` rather than `@serwist/next`, meaning that even if you don't use Next.js, you can still use the feature :) Hope this helps!

**Reply by ajayvignesh01** [reply]
Can you explain the revision please. Confused on it's purpose and what to set it as.

**Reply by DuCanhGH** [reply]
@0x-Legend the revision's purpose is to help `@serwist/precaching.precacheAndRoute` does its job of precaching the fallback URLs without fault. You can set it to anything random. In the example above, it is created using `crypto.randomUUID()`, which has been implemented in all major browsers as of now:
![image](https://github.com/serwist/serwist/assets/75556609/dce37839-9416-4e07-a5fe-ad59e6f46ece)
You can use anything you like though. For example, the package `uuid` may work as an alternative.
